### PR TITLE
Feat: add pair table model to pytorch

### PR DESCRIPTION
**Pull request by anyangml**
Migrated from this [PR](https://github.com/dptech-corp/deepmd-pytorch/pull/174). This is to reimplement the PairTab Model in Pytorch.

Notes:

1. different from the tensorflow version, the pytorch version abstracts away all the post energy conversion operations (force, virial).
2. the pytorch version implements additional checks on user provided table against the `rcut`, linear extrapolation will be performed when necessary. Check `_check_table_upper_boundary` method for details.

### CODE REVIEW

File: deepmd/pt/model/model/pair_tab.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+                [0.01  0.8   1.6   2.4  ]
+                [0.015 0.    1.    1.5  ]]
+
+        rcut = 0.022
+
+        new_table = [[0.005 1.    2.    3.   ]
+                    [0.01  0.8   1.6   2.4  ]
+                    [0.015 0.    1.    1.5  ]
+                    [0.02  0.    0.5   0.75 ]
+                    [0.025 0.    0.    0.   ]]
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/model/pair_tab.py:164]: We requires that the table strictly decays to zero at rcut, thus if the cut-off is 0.022, then the expected values are 
```
[0.02  0.    0.    0. ]
[0.025 0.    0.    0. ]]
```
otherwise we have a small but non-zero value at the rcut.

Comment (anyangml) [reply deepmd/pt/model/model/pair_tab.py:164]: i see.

File: deepmd/pt/model/model/pair_tab.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        atomic_energy = 0.5 * torch.sum(
+            torch.where(
+                nlist != -1, raw_atomic_energy, torch.zeros_like(raw_atomic_energy)
+            ),
+            dim=-1,
+        )
+
+        return {"energy": atomic_energy}
+
+    def _check_table_upper_boundary(self):
```
Comment (njzjz) [inline deepmd/pt/model/model/pair_tab.py:143]: Why not directly add to the `PairTab` class? This method just modifies the data but is not related to PT. TF can also use it.

File: deepmd/pt/model/model/pair_tab.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+                    [
+                        np.linspace(start, 0, rcut_idx - upper_idx + 1)
+                        for start in upper_val
+                    ]
+                ).T
+                raw_data = np.concatenate((raw_data[:-1, :], pad_linear), axis=0)
+
+        # over writing file with padding if applicable.
+        with open(self.tab_file, "wb") as f:
+            np.savetxt(f, raw_data)
```
Comment (njzjz) [inline deepmd/pt/model/model/pair_tab.py:234]: Overwriting the user's provided file is not a good idea.

Comment (anyangml) [reply deepmd/pt/model/model/pair_tab.py:231]: three options here:

1. overwriting the file: I agree, this may not be the best practice, even though all the original information is still kept in the file.
2. write to a new file
3. change the `PairTab` and allow both `file path` and `np.array` as input. With this, we can easily add test cases, without writing new files.

File: deepmd/pt/model/model/pair_tab.py (hunk 2 above)
Comment (anyangml) [reply deepmd/pt/model/model/pair_tab.py:143]: sure, if that's what we want.

File: deepmd/pt/model/model/pair_tab.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+        rcut = 0.031
+
+        new_table = [[0.005 1.    2.    3.   ]
+                    [0.01  0.8   1.6   2.4  ]
+                    [0.015 0.5   1.    1.5  ]
+                    [0.02  0.25  0.4   0.75 ]
+                    [0.025 0.    0.1   0.   ]
+                    [0.03  0.    0.    0.   ]
+                    [0.035 0.    0.    0.   ]]
```
Comment (anyangml) [inline deepmd/pt/model/model/pair_tab.py:183]: @wanghan-iapcm I wonder for those padding zeros, when creating cubic spline, will that result in negative energy values?

Comment (anyangml) [reply deepmd/pt/model/model/pair_tab.py:183]: the easiest fix for this would be to get the number of zero padding rows from the updated table, and overwrite the cubic spline coefficients before calculating energy.

File: deepmd/pt/model/model/pair_tab.py (hunk 3 above)
Comment (anyangml) [reply deepmd/pt/model/model/pair_tab.py:231]: @njzjz @wanghan-iapcm I lean towards option 3

Comment (njzjz) [reply deepmd/pt/model/model/pair_tab.py:231]: It should not be the input of `PairTab`. Instead, it should be internal within `PairTab`.

File: deepmd/pt/model/model/pair_tab.py (hunk 4 above)
Comment (wanghan-iapcm) [reply deepmd/pt/model/model/pair_tab.py:183]: > @wanghan-iapcm It seems when padding with these zeros, it may cause the cubic spline go below zero and result in negative energy values.

I agree. it is a numerical issue.  The solution would be construct a smooth spline (continuous up to the 2nd order at table_up and rcut) between table_up and rcut.

File: deepmd/utils/pair_tab.py [hunk 5]
```diff
@@ -73,11 +163,18 @@ def _make_data(self):
...
-                        -3 * vv[ii] + 3 * vv[ii + 1] - 2 * dd[ii] - dd[ii + 1]
+                        (-3 * vv[ii] + 3 * vv[ii + 1] - 2 * dd[ii] - dd[ii + 1])
+                        if vv[ii] != 0
+                        else 0
                     )
-                    dtmp[ii * 4 + 2] = dd[ii]
+                    dtmp[ii * 4 + 2] = dd[ii] if vv[ii] != 0 else 0
                     dtmp[ii * 4 + 3] = vv[ii]
                 data[
                     (t0 * self.ntypes + t1) * stride : (t0 * self.ntypes + t1) * stride
```
Comment (anyangml) [inline deepmd/utils/pair_tab.py:180]: if the table value at the beginning of the given spline is already `0` then, we set all 4 coefficients to `0`, to prevent negative energy.

File: deepmd/pt/model/model/pair_tab.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+        uu -= idx
+
+        final_coef = self._extract_spline_coefficient(i_type, j_type, idx)
+
+        # here we need to do postprocess to overwrite coefficients when table values are not zero at rcut in linear extrapolation.
+        if self.tab.rmax < self.rcut:
+            post_mask = rr >= self.rcut
+            final_coef[post_mask] = torch.zeros(4)
+
```
Comment (anyangml) [inline deepmd/pt/model/model/pair_tab.py:198]: this differs from the tensorflow implementation, and may lead to inconsistent results. this is to handle a specific edge case, where rcut is slightly larger than the table upper boundary. for example, the table upper boundary is `0.2`, and the rcut is `0.22`, if the space between grids (hh) is `0.05`, in the current implementation, the table values do not decay to `0` at `0.22`, but at `0.25`. So for values between `0.2` and `0.22`, the values are calculated based on linear extrapolation in range 0.2~0.25, and the values at and beyond 0.22 are forced to `0`

File: deepmd/utils/pair_tab.py (hunk 5 above)
Comment (anyangml) [reply deepmd/utils/pair_tab.py:180]: We decided to ignore this case, and allow negative energy.

File: deepmd/pt/model/model/pair_tab.py (hunk 6 above)
Comment (anyangml) [reply deepmd/pt/model/model/pair_tab.py:198]: Note, this logic is not implemented in the Tensorflow version, need to add to the c++ operator

File: source/tests/tf/test_pairtab_preprocess.py
Comment (njzjz) [inline source/tests/tf/test_pairtab_preprocess.py:1]: This file should move to `common` directory

**Comment by njzjz** [comment]
Does this PR fix #1895 or #1973?

**Comment by anyangml** [comment]
> Does this PR fix #1895 or #1973?

It's related to these issues, and partially solves them. The extrapolation feature added in this PR only handles the case where `rcut` is larger than the table upper boundary, it will calculate atomic energy with cubic spline extrapolation for atoms in the range between table upper boundary and `rcut`.

**Review by njzjz — APPROVED** [review]

File: deepmd/pt/model/model/pair_tab.py [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+        ener = self._calcualte_ener(table_coef, uu)
+
+        if self.tab.rmax <= self.rcut:
+            # here we need to overwrite energy to zero beyond rcut.
+            mask_beyond_rcut = rr > self.rcut
+            ener[mask_beyond_rcut] = 0
+
+            # here we use smooth extrapolation to replace linear extrapolation.
+            extrapolation = self._extrapolate_rmax_rcut()
+            if extrapolation is not None:
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/model/pair_tab.py:205]: shouldn't the extrapolation implemented in the table preparation stage and made backend independent? 
https://github.com/deepmodeling/deepmd-kit/blob/2ba031889824120422b19e7b1b2290d170ade850/deepmd/utils/pair_tab.py#L141-L148

Comment (anyangml) [reply deepmd/pt/model/model/pair_tab.py:205]: The `linspace` in the current implementation is to maintain the grid space. The smooth extrapolation is overwriting the linear extrapolation. 
To use smooth extrapolation in `table preparation stage`, we need to use cubic spline to calculate those points, and then in the `_make_data` method use cubic spline again to calculate the coefficient.

**Review by wanghan-iapcm — APPROVED** [review]
